src/core/connection.c

```diff
@@ -1441,13 +1441,15 @@ QuicConnTryClose(
     }
 
     if (!ClosedRemotely) {
-
         if ((Flags & QUIC_CLOSE_APPLICATION) &&
-            Connection->Crypto.TlsState.WriteKey < QUIC_PACKET_KEY_1_RTT) {
+            !Connection->State.HandshakeConfirmed) {
             //
             // Application close can only happen if we are using 1-RTT keys.
-            // Otherwise we have to send "user_canceled" TLS error code as a
-            // connection close. Overwrite all application provided parameters.
+            // But until the handshake is confirmed, we can't be sure if the
+            // peer is able to process the close frame.
+            //
+            // Send "user_canceled" TLS error code as a connection close instead.
+            // Overwrite all application provided parameters.
             //
```

Comment:
Open question: This change makes the behavior symmetrical, it is possible for the server to send Application CONNECTION_CLOSE because client has the keys to read it.

Reply:
The condition was originally introduced by https://github.com/microsoft/msquic/pull/3186

Reply:
I would prefer to minimize behavior change, and limit this to client side only.